Support new schema (resolves LOAD-18, LOAD-17, LOAD-20, LOAD-21)

jessebrennan opened the pull request:
_Note: I realize there should be only one issue per PR. most of those are sub-tasks of the main issue LOAD-18, and the last (LOAD-17) is essentially a test and should be included with this PR_

This PR adds support for loading data with the new metadata schema defined in [shemas.py](https://github.com/DataBiosphere/cgp-dss-data-loader/compare/feature/new-schema#diff-ab51b0be30526ca96aa5c76078e556b7).

**What's new:**
 - [a revised README](https://github.com/DataBiosphere/cgp-dss-data-loader/compare/feature/new-schema#diff-04c6e90faac2675aa89e2176d2eec7d8) explaining how to run the loader with the new format and how its different from loading from the old format
 - Tests that run against the output of the new [**newt-transformer**](https://github.com/jessebrennan/newt-transformer#transforming-data-from-sheepdog-exporter) (LOAD-20)
 - Test for the metadata.json that gets loaded to make sure that it's formatted according to the schema we give to the dss (LOAD-17)
 - The new loader that can understand the new format that is output from **newt-transformer** (LOAD-18, LOAD-21)

**What still needs to be done:**
 

- [ ] I haven't yet tried to upload the 107 using the new format. I was waiting for input from @mikebaumann before doing this. Since the new 'standard' loader passes tests, I don't imagine this will be an issue however.
- [ ] Possibly (depending on input of @david4096 and @mikebaumann) remove the old gen3 loader. I think we should keep it around for at least another iteration as a backup.

┆Issue is synchronized with this [JIRA Task](https://ucsc-cgl.atlassian.net/browse/LOAD-24)
┆Project Name: CGP-Data Loader
┆Issue Number: LOAD-24

jessebrennan committed:
Make test for loader more abstract

jessebrennan committed:
Add (hopefully) working tests to the standard loader (resolves LOAD-20)

Tests should now fail on Travis because the standard loader isn't implemented yet. See next commit.

Tests do basically the same thing as the gen3 tests, but adapted to the new metadata format.

jessebrennan committed:
Add test that metadata matches the DSS schema (resolves LOAD-17)

Before, the schema was of basically the wrong thing. The dss didn't complain before, but this new test should prevent this mistake in the future.

jessebrennan committed:
Add loader support for standard format (resolves LOAD-21, LOAD-18)

jessebrennan committed:
Add new loading instructions to README

jessebrennan committed:
Make tests more verbose

jessebrennan committed:
Add better logging

jessebrennan commented:
Well now we have a logging system set up and the mysteriously failing Travis builds have mysteriously resumed passing. Looking at the log from one of the failing builds, [68](https://travis-ci.org/DataBiosphere/cgp-dss-data-loader/builds/394781688) there was a socket error which makes me think this was maybe a network error on Travis?

Perhaps this indicates that the retry mechanism used for the loader needs some improvement. @mikebaumann can you weigh in?

mikebaumann commented on loader/standard_loader.py:
```diff
@@ -1,11 +1,62 @@
+
+    @staticmethod
+    def _get_file_ids(file_guid: str):
+        return file_guid.split('/')[1]
```
We should get @david4096's thoughts on this, yet I think going forward we are likely to see different identifier formats, especially as we try to interoperate with other (non Team Calcium) stacks and data sources.
The code currently appears to assume the id format is of the form:
    `dg.4503/00e6cfa9-a183-42f6-bb44-b70347106bbe`
Yet I think it is reasonable/appropriate for the current standard loader to handle id formats such as:
    `dos://dg.4503/00e6cfa9-a183-42f6-bb44-b70347106bbe`
    `00e6cfa9-a183-42f6-bb44-b70347106bbe`
which would fail with the code as currently written.
As this point, I am willing to make the assumption that the id contains a UUID in it somewhere, so one option would be a regular expression that identifies the UUID in the id and extracts it.
At some point, as we interoperate with other stacks or data sources we may have input data in which the identifiers don't contain a UUID, but I think we can address that when we come to it, and likely address that in the transformer when we do.

mikebaumann commented on loader/standard_loader.py:
```diff
@@ -1,11 +1,62 @@
+        return {url_dict['url'] for url_dict in file_info['urls']}
+
+    @staticmethod
+    def _get_file_ids(file_guid: str):
```
I think this method may be better named `_get_file_uuid` because is what we need to provide when loading into the DSS.
This method should also verify that the result of this function is indeed a valid UUID pattern (using a regex) and raise an exception if it is not.

mikebaumann commented on loader/standard_loader.py:
```diff
@@ -1,11 +1,62 @@
+        return file_guid.split('/')[1]
+
+    def _load_bundle(self, bundle: dict):
+        bundle_uuid, metadata_dict, data_objects = self._parse_bundle(bundle)
```
This method should log information about the bundle being loaded, both so there is some visible indication of progress when loading large data sets, and if there is an error with loading a bundle, we know which bundle this occurred in.

mikebaumann commented on loader/standard_loader.py:
```diff
@@ -1,11 +1,62 @@
+
+    def load_all_bundles(self, input_json: list):
+        for bundle in input_json:
+            self._load_bundle(bundle)
```
We need to consider what the behavior of the data loader is when there is an error loading the bundle.
Should it stop loading at the first bundle that encounters an error, or should continue on and try to load all valid bundles? Arguments can be made either way, although I am inclined towards the later, in which case the a try/except should be placed around `self._load_bundle`.

mikebaumann replied:
This should log a summary including:
- the total number of bundles that were (attempted to be) loaded, the number of successfully
- the total number of bundles that failed to load
- a list of the ids of the bundles that failed to load

jessebrennan replied:
It's a great idea to include all of this information! It will be vital when doing large uploads. 

Dumping all of that info to stdout by default seems a little unwieldy to me. How to you feel about having a command line option to dump that info to files instead?

mikebaumann commented on loader/standard_loader.py:
```diff
@@ -1,11 +1,62 @@
+        # load bundle
+        self.dss_uploader.load_bundle(file_info_list, bundle_uuid)
+
+    def load_all_bundles(self, input_json: list):
```
When loading large data sets (e.g. the TOPMed 12k) it would be useful to log the current count of bundles loaded so far (1, 2, 3 ... 9786, 9787, ... 12000) to have an idea of how the load is progressing and how many bundles remain to be loaded.

jessebrennan replied:
This is a good idea and I'll implement it now, but will become more difficult when bundles are loaded in parallel.

mikebaumann commented on loader/standard_loader.py:
```diff
@@ -1,11 +1,62 @@
+            file_uuid = self._get_file_ids(file_guid)
+            file_uuid, file_version, filename = \
+                self.dss_uploader.upload_cloud_file_by_reference(file_info['name'],
+                                                                 # use did for uuid for now. will probably have to
```
This comment seems obsolete at this point and should be removed, I think.

mikebaumann commented on loader/schemas.py:
```diff
@@ -0,0 +1,232 @@
+
+# These schemas are based on the DOS schemas with minor modifications. See
+# https://github.com/ga4gh/data-object-service-schemas/blob/master/openapi/data_object_service.swagger.yaml#L594
```
A brief list of the modifications/differences in this schema as compared to the standard DOS schema would be useful here, including a "one-line" description of why where appropriate.

mikebaumann commented on README.md:
```diff
@@ -43,12 +43,30 @@ Run:
-   
-   `python transformer/gen3_transformer.py /path/to/topmed_public.json --output-json transformed-topmed-public.json` 
-   
+3. Now we need to transform the data. We can transform to the  outdated gen3 format, or two the new standard format.
```
Change `two` to `to`

mikebaumann commented on loader/schemas.py:
```diff
@@ -0,0 +1,232 @@
+        'id': {
+            'type': 'string',
+            'description': 'REQUIRED\n'
+                           'An identifier unique to this Data Object.'
```
Should add to the description that the id is required to contain a valid UUID.

mikebaumann commented on loader/schemas.py:
```diff
@@ -0,0 +1,232 @@
+        'name': {
+            'type': 'string',
+            'description': 'OPTIONAL\n'
+                           'A string that can be optionally used to name a Data Object.'
```
Is the name really optional for the data loader? I suspect not.

mikebaumann commented on loader/standard_loader.py:
```diff
@@ -1,11 +1,74 @@
+            file_info = data_objects[file_guid]
+            filename = file_info['name']
+            file_uuid = self._get_file_ids(file_guid)
+            file_version = file_info['updated']
```
As I recall, we discussed using the file `updated` value if it is provided (shown as optional in the schema), otherwise use the `created` value (which is required according to the schema provided).
This code doesn't do that, yet should.

jessebrennan replied:
Whoops! nice catch.

mikebaumann commented on loader/schemas.py:
```diff
@@ -0,0 +1,232 @@
+# These schemas are based on the DOS schemas with minor modifications. See
+# https://github.com/ga4gh/data-object-service-schemas/blob/master/openapi/data_object_service.swagger.yaml#L594
+
+data_object_schema = {
```
I think this schema should provide an accurate and concise description of what the transformers must produce as output for the data to be successfully loaded into the DSS. Some of the fields identified by the DOS schema as optional are actually required for loading into the DSS, and some fields identified by the DOS schema as required are actually optional for loading into the DSS, and in fact, not even used by the standard loader.
@david4096, what are your thoughts on this?

jessebrennan replied:
~~The idea is that this schema should only reflect what is required for loading. Anything non-conforming should be changed (as with the 'name' field that you pointed out earlier).~~

I take this back. I'm undecided. We can have a bare bones schema like you suggest, but I think it's worth considering, are we ever even gonna want to have data that doesn't have checksums for data objects even though it could be uploaded (for example)?

mikebaumann commented on loader/standard_loader.py:
```diff
@@ -1,11 +1,74 @@
+            logger.debug(f'Uploading data file: {filename} with uuid:version {file_uuid}:{file_version}...')
+            file_uuid, file_version, filename = \
+                self.dss_uploader.upload_cloud_file_by_reference(filename,
+                                                                 # use did for uuid for now. will probably have to
```
I think this comment is obsolete and should be removed.

mikebaumann commented on scripts/cgp_data_loader.py:
```diff
@@ -38,6 +38,9 @@ def main(argv):
                         help="Bucket to stage local files for uploading to DSS")
+    parser.add_argument("-l", "--log", dest="logLevel",
+                        choices=['DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL'],
+                        default="WARNING", help="Set the logging level")
```
I think the default should INFO, as in general we want a reasonable amount of informational output as large loads progress, etc.

jessebrennan replied:
I was planning to add a `--verbose` flag that would set logLevel to INFO and use `verbose` in the example code. I'm happy to change it either way.

mikebaumann commented on tests/test_data/gen3_sample_input_standard_metadata.json:
```diff
@@ -0,0 +1,218 @@
+[
```
We should have a test with input that contains more than one bundle (e.g. two bundles)

jessebrennan replied:
Agreed.

mikebaumann commented on tests/test_data/gen3_sample_input_standard_metadata.json:
```diff
@@ -0,0 +1,218 @@
+[
+  {
```
We should have a test with input that contains only fields actually required for loading into the DSS (and indicated as such in the schema).

mikebaumann commented on tests/test_standard_input_json.py:
```diff
@@ -1,23 +1,169 @@
+            for r in search_results['results']:
+                response = requests.get(r['bundle_url'])
+                returned_json = response.json()
+                for f in returned_json['bundle']['files']:
```
This test really should verify that the number of data files loaded into the bundle matches those specified in the input file. As it stands now, this test would pass if no data files were loaded, or if a subset of the data files in the input were loaded (and others were not).
I think the test should also verify that a `metadata.json` file is present, as right now the test would pass even if it were not. Ideally, the test should verify that the metadata.json contains the expected metadata based on the input.

mikebaumann commented on loader/schemas.py:
```diff
@@ -0,0 +1,232 @@
+
+data_object_schema = {
+    'type': 'object',
+    'required': ['id', 'size', 'created', 'checksums', 'urls'],
```
The description comments of "REQUIRED" or "OPTIONAL" should match what is currently required for the loader, as identified here in the `required` statement.
Rather than trying to keep the `required` values and descriptions consistent, it may be simplest/cleanest to just use the `required` values and remove the "REQUIRED" or "OPTIONAL" text from the descriptions.

The values `size` and `checksums` are not actually used by the loader today, although I think it is reasonable/good to keep them as required in the schema as they may potentially be used by the loader in the future.

mikebaumann commented on loader/schemas.py:
```diff
@@ -0,0 +1,232 @@
+data_bundle = {
+    "type": "object",
+    # TODO: required, but we don't have checksums for bundles
+    'required': ['data_object_ids', 'created', 'updated', 'version'],  # , 'checksums'],
```
Ditto comment above regarding consistency between the values of `required` here and the use of "OPTIONAL" and "REQUIRED" in the descriptions.

mikebaumann commented on tests/test_standard_input_json.py:
```diff
@@ -1,23 +1,169 @@
-        cls.project_path = Path(__file__).parents[1]
+        cls.test_file = TEST_DATA_PATH / 'gen3_sample_input_standard_metadata.json'
+
+    def test_data_matches_schema(self):
```
Ideally there should also be test(s) for the case in which the input data file does not contain all of the required fields. What happens when one (or more) bundles in the input file do not contain all of the required fields?

mikebaumann commented on tests/test_standard_input_json.py:
```diff
@@ -1,23 +1,169 @@
+                '--staging-bucket',
+                f'{self.staging_bucket}',
+                '--log',
+                'DEBUG',
```
In general, tests should run at the default log level. An alternate log level should not be specified as part of the standard test input parameters.

jessebrennan replied:
I set it like this because originally I was trying to debug Travis and needed the log. I found it very helpful as well when trying to debug failing tests with Toil. I think having the full log for unit tests can be valuable for these reasons. Can you convince me why it should be default?

mikebaumann reviewed (changes requested):
Great job, Jesse!
This is nice, clean easy to read/understand code.
Please address the comments provided and let me know when it is ready for re-review.
Thanks!

jessebrennan committed:
Make loading more robust and improve logging

jessebrennan committed:
Tidy up schemas

changes:
 - The OPTIONAL|REQUIRED tag was removed from the description
 - The differences from DOS are made explicit
 - Name is made required

Another note:
The fields required here is a superset of the fields required to load. Whether this superset should be reduced is a question yet to be answered.

jessebrennan committed:
Implement review suggestions, + some cleanup

This should include most of Michael's suggestions. 

Yet to be addressed were test improvements. A TODO was added to the tests with a list of more tests that need to be added.

jessebrennan commented:
Okay @mikebaumann, I've implemented most of your suggestions. You found a lot of important mistakes and gave me a lot of ideas for improvements! I've responded to some of your comments above, but I'll summarize everything again.

1. I haven't implemented most of your suggestions for improved tests. I thought of some new ones myself while doing your suggestions. I compiled a list and propose to postpone these test improvements for another PR since it seems like it will be a good amount of work.

2. I set log level to INFO by default as per your suggestion but propose keeping it at WARNING but adding a `--verbose` option that switches to INFO. I'm not too attached either way.

3. You suggested making tests run the default log level (INFO). I think it can be really handy having log DEBUG for tests. I set up logging in the first place in order to debug Travis, and in general when tests fail it's useful to be able to see the full log without having to run modified tests again. Can you convince me why it's worth it to use the default log level for tests?

4. The schemas require some fields that are not required for loading (but all fields required for loading are required by the schemas). Should this be changed? I think this could be discussed in person.

5. The way I've it set up now, when a bundle fails to load, its information is logged real time. You suggested making a summary of all of the bundles that fail and logging it at the end. I think that's a great idea, but was considering adding a command line option to output the the summary to a file. What do you think?

mikebaumann commented on loader/standard_loader.py:
```diff
@@ -1,11 +1,135 @@
+
+    @staticmethod
+    def _get_file_uuid(file_guid: str):
+        uuid_regex = re.compile('[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}')
```
This line could be moved to the class level, so avoid unnecessary recompilation of the regex each time this method is called.

jessebrennan replied:
I'm going to ignore this for now since:
>Note The compiled versions of the most recent patterns passed to re.compile() and the module-level matching functions are cached, so programs that use only a few regular expressions at a time needn’t worry about compiling regular expressions.

https://docs.python.org/3/library/re.html#re.compile

jessebrennan replied:
Never mind, I just fixed it anyway.

mikebaumann reviewed (approved):
This looks great, Jesse! Thank you!
I believe I can convince you that tests should not run at debug log level by default for several important reasons. Yet, I accept this as it stands (for now).
I did add one small comment which could be addressed, yet not required.

jessebrennan committed:
Prevent recompiling of regex
